src/connection.c

```diff
@@ -330,3 +328,24 @@ ConnectionGetOptionValue(PGconn *connection, char *optionKeyword)
 
 	return optionValue;
 }
+
+ShardId DoForAllShards(List* shardConnections, ShardAction action, void* arg)
```

Comment:
Yep. It's reasonably simple code at this point, so the redundancy is small. My concern that is the failure semantics are different depending on how this function is called. If a prepare fails we should stop immediately, and if an abort fails we should keep going. It's not clear from the DoForAllShards function that it does this to an uninformed reader, since he wouldn't don't know the return values of the functions, nor is it clear from looking at the call to DoForAllShards, nor is it clear from looking at PgCopyPrepareTransaction or PgCopyEndTransaction individually. He will have to look at 3-5 different places and think about how they tie together before knowing what happens in case of failure. By making the behaviour explicit in separate functions the behaviour will be obvious and will only take a few seconds to understand, which makes it much easier to maintain the code even if it is a few more lines.